src/lib/protocols/credentials/__tests__/CredentialService.test.ts

```diff
@@ -0,0 +1,184 @@
+    await wallet.init();
+  });
+
+  afterAll(async () => {
+    await wallet.close();
+    await wallet.delete();
+  });
+
+  describe('createCredentialOffer', () => {
+    beforeEach(() => {
+      storageService = new StubStorageService();
+      credentialRepository = new Repository<CredentialRecord>(CredentialRecord, storageService);
+      credentialService = new CredentialService(wallet, credentialRepository);
+    });
+
+    test('returns credential offer message', async () => {
+      const credentialOffer = await credentialService.createCredentialOffer({
+        credDefId: 'Th7MpTaRZVRYnPiabds81Y:3:CL:17:TAG',
+        comment: 'some comment',
+        preview,
+      });
+
+      expect(credentialOffer.toJSON()).toEqual(
+        expect.objectContaining({
+          '@id': expect.any(String),
+          '@type': 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/issue-credential/1.0/offer-credential',
+          comment: 'some comment',
+          credential_preview: expect.any(Object),
+          'offers~attach': [
+            {
+              '@id': expect.any(String),
+              'mime-type': 'application/json',
+              data: {
+                base64: expect.any(String),
+              },
+            },
+          ],
+        })
+      );
+
+      // @ts-ignore
+      expect(credentialOffer.toJSON().credential_preview).toEqual({
+        type: 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/issue-credential/1.0/credential-preview',
+        attributes: [
+          {
+            name: 'name',
+            mimeType: 'text/plain',
+            value: 'John',
+          },
+          {
+            name: 'age',
+            mimeType: 'text/plain',
+            value: '99',
+          },
+        ],
+      });
+    });
+
+    test('creates credential in OFFER_SENT state', async () => {
+      const credentialOffer = await credentialService.createCredentialOffer({
+        credDefId: 'Th7MpTaRZVRYnPiabds81Y:3:CL:17:TAG',
+        comment: 'some comment',
+        preview,
+      });
+      const [firstCredential] = await credentialService.getAll();
+
+      expect(firstCredential).toEqual(
+        expect.objectContaining({
+          createdAt: expect.any(Number),
+          id: expect.any(String),
+          offer: credentialOffer,
+          tags: {},
+          type: 'CredentialRecord',
+          state: CredentialState.OfferSent,
+        })
+      );
+    });
+
+    test(`emits stateChange event with ${CredentialState.OfferSent}`, async () => {
+      const eventListenerMock = jest.fn();
+      credentialService.on(EventType.StateChanged, eventListenerMock);
+
+      await credentialService.createCredentialOffer({
+        credDefId: 'Th7MpTaRZVRYnPiabds81Y:3:CL:17:TAG',
+        comment: 'some comment',
+        preview,
+      });
+
+      expect(eventListenerMock).toHaveBeenCalledWith(
+        expect.objectContaining({
+          newState: CredentialState.OfferSent,
+          credentialId: expect.any(String),
+        })
```

Comment:
I'm still not sure what should be a convention for content returned by an event emitter. I'm returning `newState` and `credentialId` here, but maybe it would be better to return `credential` record itself and something what would highlight state change `{ from: offer-received, to: request-sent }`

Reply:
I think it would be useful to return the whole credential record. I also think it would be useful to get the previous state.

We could also use `newState` and `prevState`, but I kinda like `from` and `to`

Reply:
I think previous can still be useful. What if i receive a credential proposal and don't have the credential in memory at the moment. The proposal-received state could come from either a new proposal, or an already existing credential exchange that needs renegotiation.

Reply:
Hmm, good point 🤔